### PR TITLE
Archer - fix users specification error

### PR DESCRIPTION
**Pull request by guykeller**
## Status 
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/20591

## Description
Specifying type 8 fields with users while creating/updating a record used to fail.
It does not anymore.
Solution tested with a customer.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

### CODE REVIEW

File: Integrations/integration-ArcherRSA.yml [hunk 1]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+    function getGroupIDFromGroupName(groupsDataFromArcher, userSuppliedGroupName) {
         var desiredGroupID = -1;
         for (var i=0; i<groupsDataFromArcher.length; i++){
-            var currentGroupData = groupsDataFromArcher[i].RequestedObject;
-            var currentGroupName = currentGroupData.Name;
-            if (currentGroupName.trim() === desiredGroupName.trim()){
-                desiredGroupID = currentGroupData.Id;
+            var currentArcherGroupData = groupsDataFromArcher[i].RequestedObject;
+            var currentArcherGroupName = currentArcherGroupData.Name;
+            if (currentArcherGroupName.trim() === userSuppliedGroupName.trim()){
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]: better to compare strings which are in the same case, so lowercase

File: Integrations/integration-ArcherRSA.yml [hunk 2]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+                desiredGroupID = currentArcherGroupData.Id;
             }
         }
         return desiredGroupID;
     }
 
+
+    function getUserIDFromUserName(allUsersArcherData, userSuppliedUserName) {
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]: same comments

File: Integrations/integration-ArcherRSA.yml [hunk 3]
```diff
@@ -2387,46 +2387,85 @@ script:
...
 
-    function getGroupIDFromGroupName(groupsDataFromArcher, desiredGroupName) {
+    function getGroupIDFromGroupName(groupsDataFromArcher, userSuppliedGroupName) {
         var desiredGroupID = -1;
         for (var i=0; i<groupsDataFromArcher.length; i++){
-            var currentGroupData = groupsDataFromArcher[i].RequestedObject;
-            var currentGroupName = currentGroupData.Name;
-            if (currentGroupName.trim() === desiredGroupName.trim()){
-                desiredGroupID = currentGroupData.Id;
+            var currentArcherGroupData = groupsDataFromArcher[i].RequestedObject;
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]: are you sure `RequestedObject` will always be in `groupsDataFromArcher[i]` ?

File: Integrations/integration-ArcherRSA.yml [hunk 4]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+            var currentArcherGroupName = currentArcherGroupData.Name;
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]: are you sure `Name ` will always be in `currentArcherGroupData` ?

File: Integrations/integration-ArcherRSA.yml [hunk 5]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+    }
+
+
     function getGroupData() {
         var groupsData = sendRestRequest("/api/core/system/group", 'archer-create-record', '', true)
         return groupsData;
     }
 
+    function getUserData() {
+        var userData = sendRestRequest("/api/core/system/user", 'archer-create-record', '', true)
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
        var userData = sendRestRequest("/api/core/system/user", 'archer-create-record', '', true);
```

File: Integrations/integration-ArcherRSA.yml [hunk 6]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+            if (currentArcherUserName.trim() === userSuppliedUserName.trim()){
+                desiredUserID = currentArcherUserData.Id;
+            }
+        }
+        return desiredUserID;
+    }
+
+
     function getGroupData() {
         var groupsData = sendRestRequest("/api/core/system/group", 'archer-create-record', '', true)
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
        var groupsData = sendRestRequest("/api/core/system/group", 'archer-create-record', '', true);
```

File: Integrations/integration-ArcherRSA.yml [hunk 7]
```diff
@@ -2387,46 +2387,85 @@ script:
...
         var desiredGroupID = -1;
         for (var i=0; i<groupsDataFromArcher.length; i++){
-            var currentGroupData = groupsDataFromArcher[i].RequestedObject;
-            var currentGroupName = currentGroupData.Name;
-            if (currentGroupName.trim() === desiredGroupName.trim()){
-                desiredGroupID = currentGroupData.Id;
+            var currentArcherGroupData = groupsDataFromArcher[i].RequestedObject;
+            var currentArcherGroupName = currentArcherGroupData.Name;
+            if (currentArcherGroupName.trim() === userSuppliedGroupName.trim()){
+                desiredGroupID = currentArcherGroupData.Id;
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]: are you sure` Id` will always be in `currentArcherGroupData` ?

File: Integrations/integration-ArcherRSA.yml [hunk 8]
```diff
@@ -2387,46 +2387,85 @@ script:
...
-                    var groupsWithIDsInsteadOfNames = [];
-                    for (var groupIndex = 0; groupIndex < groupsFromUser.length; groupIndex++) {
-                        var currentGroup = groupsFromUser[groupIndex];
-                        var currentGroupData = currentGroup.RequestedObject;
-                        var currentGroupName = currentGroupData.Name;
-                        var currentGroupID = getGroupIDFromGroupName(allGroupsArcherData, currentGroupName)
-                        groupsWithIDsInsteadOfNames.push(currentGroupID)
+                    var usersListFromUser = fieldValue.users;
+
+                    if (!(groupsFromUser == null)) {
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                    if (groupsFromUser)) {
```

File: Integrations/integration-ArcherRSA.yml [hunk 9]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+                            var currentGroupID = getGroupIDFromGroupName(allGroupsArcherData, currentUserSuppliedGroupName)
+
+                            groupsWithIDsInsteadOfNames.push(currentGroupID)
+                        }
+
+                        fieldValue["groups"] = groupsWithIDsInsteadOfNames;
+                        modifiedfieldsToValues[fieldName] = fieldValue;
+                    }
+
+                    if (!(usersListFromUser == null)) {
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                    if (usersListFromUser) {
```

File: Integrations/integration-ArcherRSA.yml [hunk 10]
```diff
@@ -2387,46 +2387,85 @@ script:
         })
     }
 
-    function getGroupIDFromGroupName(groupsDataFromArcher, desiredGroupName) {
+    function getGroupIDFromGroupName(groupsDataFromArcher, userSuppliedGroupName) {
         var desiredGroupID = -1;
         for (var i=0; i<groupsDataFromArcher.length; i++){
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
        for (var i = 0;  i < groupsDataFromArcher.length; i++){
```

File: Integrations/integration-ArcherRSA.yml [hunk 11]
```diff
@@ -2387,46 +2387,85 @@ script:
...
-    function replaceMultiselectNameWithID(fieldsToValues, fieldMapping) {
+    function replaceMultiselectNameWithID(fieldsToValues, createRecordMapping) {
         var modifiedfieldsToValues = fieldsToValues;
         var allGroupsArcherData = getGroupData();
+        var allUsersArcherData = getUserData();
         for (var fieldName in fieldsToValues) {
-            if (fieldMapping[fieldName] !== undefined) {
-                var fieldValue = fieldsToValue[fieldName];
-                var fieldNameMapping = fieldMapping[fieldName];
+            if (createRecordMapping[fieldName] !== undefined) {
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
            if (createRecordMapping[fieldName]) {
```

File: Integrations/integration-ArcherRSA.yml [hunk 12]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+                    if (!(groupsFromUser == null)) {
+                        var groupsWithIDsInsteadOfNames = [];
+                        for (var groupIndex = 0; groupIndex < groupsFromUser.length; groupIndex++) {
+                            var currentUserSuppliedGroupName = groupsFromUser[groupIndex];
+                            var currentGroupID = getGroupIDFromGroupName(allGroupsArcherData, currentUserSuppliedGroupName)
+
+                            groupsWithIDsInsteadOfNames.push(currentGroupID)
+                        }
+
+                        fieldValue["groups"] = groupsWithIDsInsteadOfNames;
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                        fieldValue['groups'] = groupsWithIDsInsteadOfNames;
```

File: Integrations/integration-ArcherRSA.yml [hunk 13]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+                        modifiedfieldsToValues[fieldName] = fieldValue;
+                    }
+
+                    if (!(usersListFromUser == null)) {
+                        var usersWithIDsInsteadOfNames = [];
+                        for (var userIndex = 0; userIndex < usersListFromUser.length; userIndex++) {
+                            var currentUserSuppliedUserName = usersListFromUser[userIndex];
+                            var currentUserID = getUserIDFromUserName(allUsersArcherData, currentUserSuppliedUserName)
+
+                            usersWithIDsInsteadOfNames.push(currentUserID)
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                            usersWithIDsInsteadOfNames.push(currentUserID);
```

File: Integrations/integration-ArcherRSA.yml [hunk 14]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+
+                        fieldValue["groups"] = groupsWithIDsInsteadOfNames;
+                        modifiedfieldsToValues[fieldName] = fieldValue;
+                    }
+
+                    if (!(usersListFromUser == null)) {
+                        var usersWithIDsInsteadOfNames = [];
+                        for (var userIndex = 0; userIndex < usersListFromUser.length; userIndex++) {
+                            var currentUserSuppliedUserName = usersListFromUser[userIndex];
+                            var currentUserID = getUserIDFromUserName(allUsersArcherData, currentUserSuppliedUserName)
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                            var currentUserID = getUserIDFromUserName(allUsersArcherData, currentUserSuppliedUserName);
```

File: Integrations/integration-ArcherRSA.yml [hunk 15]
```diff
@@ -2387,46 +2387,85 @@ script:
...
+
+                            usersWithIDsInsteadOfNames.push(currentUserID)
+                        }
+
+                        fieldValue["users"] = usersWithIDsInsteadOfNames;
```
Comment (Itay4) [inline Integrations/integration-ArcherRSA.yml]:
```suggestion
                        fieldValue['users'] = usersWithIDsInsteadOfNames;
```

File: Integrations/integration-ArcherRSA.yml (hunk 1 above)
Comment (guykeller) [reply Integrations/integration-ArcherRSA.yml]: Fixed

File: Integrations/integration-ArcherRSA.yml (hunk 2 above)
Comment (guykeller) [reply Integrations/integration-ArcherRSA.yml]: same to what?

File: Integrations/integration-ArcherRSA.yml (hunk 3 above)
Comment (guykeller) [reply Integrations/integration-ArcherRSA.yml]: Yes, if group data is not null it must have the `RequestedObject` field

File: Integrations/integration-ArcherRSA.yml (hunk 4 above)
Comment (guykeller) [reply Integrations/integration-ArcherRSA.yml]: Yes, that is the structure, if there is a group it has a name

File: Integrations/integration-ArcherRSA.yml (hunk 2 above)
Comment (Itay4) [reply Integrations/integration-ArcherRSA.yml]: same to `getGroupIDFromGroupName` which is the implements the exactly same functionality

File: Integrations/integration-ArcherRSA.yml (hunk 7 above)
Comment (Itay4) [reply Integrations/integration-ArcherRSA.yml]: what about this one ?

File: Integrations/integration-ArcherRSA.yml (hunk 8 above)
Comment (Itay4) [reply Integrations/integration-ArcherRSA.yml]: what about this one ?

File: Integrations/integration-ArcherRSA.yml (hunk 9 above)
Comment (Itay4) [reply Integrations/integration-ArcherRSA.yml]: what about this one ?

File: Integrations/integration-ArcherRSA.yml (hunk 10 above)
Comment (Itay4) [reply Integrations/integration-ArcherRSA.yml]: what about this one ?

File: Integrations/integration-ArcherRSA.yml (hunk 2 above)
Comment (guykeller) [reply Integrations/integration-ArcherRSA.yml]: Fixed

**Pull request merged**